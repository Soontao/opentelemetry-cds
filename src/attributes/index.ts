
const CDS_TENANT_ID = "cds.tenant.id";
const CDS_APP_SERVICE_NAME = "cds.app.service.name";
const CDS_REQUEST_TARGET = "cds.request.target";
const CDS_REQUEST_ID = "cds.request.id";
const CDS_REQUEST_EVENT = "cds.request.event";
const CDS_SERVICE_NAME = "cds.service.name";
const CDS_SERVICE_KIND = "cds.service.kind";

const CDS_QUERY_ENTITIES = "cds.request.query.entities";
const CDS_BATCH_MULTIPART_BOUNDARY = "cds.batch.multipart.boundary";

export const CDSSemanticAttributes = {
  CDS_TENANT_ID,
  CDS_APP_SERVICE_NAME,
  CDS_REQUEST_TARGET,
  CDS_REQUEST_EVENT,
  CDS_REQUEST_ID,
  CDS_SERVICE_NAME,
  CDS_SERVICE_KIND,
  CDS_QUERY_ENTITIES,
  CDS_BATCH_MULTIPART_BOUNDARY,
};
