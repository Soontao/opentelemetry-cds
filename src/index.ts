
import { OTLPTraceExporter } from "@opentelemetry/exporter-trace-otlp-http";
import { registerInstrumentations } from "@opentelemetry/instrumentation";
import { DnsInstrumentation } from "@opentelemetry/instrumentation-dns";
import { ExpressInstrumentation } from "@opentelemetry/instrumentation-express";
import { GenericPoolInstrumentation } from "@opentelemetry/instrumentation-generic-pool";
import { HttpInstrumentation } from "@opentelemetry/instrumentation-http";
import { MySQL2Instrumentation } from "@opentelemetry/instrumentation-mysql2";
import { NetInstrumentation } from "@opentelemetry/instrumentation-net";
import { Resource } from "@opentelemetry/resources";
import { BatchSpanProcessor } from "@opentelemetry/sdk-trace-base";
import { NodeTracerProvider } from "@opentelemetry/sdk-trace-node";
import { SemanticResourceAttributes } from "@opentelemetry/semantic-conventions";
import os from "os";
import path from "path";
import process from "process";
import { CDSCompilerServiceInstrumentation } from "./instruments/CDSCompilerInstrumentation";
import { CDSNatsInstrumentation } from "./instruments/CDSNatsInstrumentation";
import { CDSServiceInstrumentation } from "./instruments/CDSServiceInstrumentation";
import { HanaInstrumentation } from "./instruments/HanaInstrumentation";
import { ODataAdapterInstrumentation } from "./instruments/ODataAdapterInstrumentation";
import { SqliteInstrumentation } from "./instruments/SqliteInstrumentation";

const provider = new NodeTracerProvider({
  resource: new Resource({
    [SemanticResourceAttributes.SERVICE_NAME]: process.env.SERVICE_NAME ?? path.basename(process.cwd()),
    [SemanticResourceAttributes.PROCESS_PID]: process.pid,
    [SemanticResourceAttributes.HOST_ARCH]: os.arch(),
    [SemanticResourceAttributes.HOST_NAME]: os.hostname(),
  }),
});

const exporter = new OTLPTraceExporter();

provider.addSpanProcessor(new BatchSpanProcessor(exporter));

provider.register();

registerInstrumentations({
  instrumentations: [
    new HttpInstrumentation(),
    new ExpressInstrumentation(),
    new CDSServiceInstrumentation(),
    new CDSCompilerServiceInstrumentation(),
    new ODataAdapterInstrumentation(),
    new GenericPoolInstrumentation(),
    new NetInstrumentation(),
    new DnsInstrumentation(),
    new MySQL2Instrumentation(),
    new SqliteInstrumentation(),
    new HanaInstrumentation(),
    new CDSNatsInstrumentation(),
  ],
});


["SIGINT", "SIGTERM"].forEach(signal => {
  process.on(signal, () => provider.shutdown().catch(console.error));
});
