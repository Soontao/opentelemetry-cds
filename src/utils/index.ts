
import * as assert from "./assert";
import * as object from "./object";

export const utils = {
  assert, object,
};

export default utils;
