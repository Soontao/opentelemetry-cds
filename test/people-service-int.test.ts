/* eslint-disable @typescript-eslint/no-var-requires */
/* eslint-disable @typescript-eslint/ban-ts-comment */
// @ts-nocheck

describe("People Service Int Test", () => {
  require("../src");
  const cds = require("@sap/cds");

  /**
   * @type {{axios:import("axios").AxiosInstance}}
   */
  const { axios } = cds.test(".").in(__dirname, "./app");

  axios.defaults.validateStatus = () => true;
  // if you enabled the basic auth for local development
  // axios.defaults.auth = { username: "user", password: 'pass' }

  const testPeopleID = "318da8b4-95be-498d-bae7-f0c6ed6516ac";

  beforeAll(() => {
    jest.spyOn(cds.db, "run");
  });


  it("should support read metadata", async () => {
    const response = await axios.get("/people/$metadata");
    expect(response.data).toMatch(/EarthPeople/);
  });

  it("should support validate the length of Name", async () => {
    const response = await axios.post("/people/EarthPeoples", { "Name": "theo" });
    expect(response.status).toBe(500);
    expect(response.data.error.message).toBe("invalid name");
  });

  it("should support create a valid EarthPeople instance", async () => {
    const response = await axios.post("/people/EarthPeoples", {
      "ID": testPeopleID,
      "Name": "theo valid"
    });
    expect(response.status).toBe(201);
    expect(response.data).toMatchObject({
      "@odata.context": "$metadata#EarthPeoples/$entity",
      "Address": null,
      "Age": 18,
      "ID": testPeopleID,
      "Name": "theo valid",
      "Weight": null,
    });
    const getResponse = await axios.get(`/people/EarthPeoples(${testPeopleID})`);
    expect(getResponse.status).toBe(200);
    expect(getResponse.data).toMatchObject({
      "@odata.context": "$metadata#EarthPeoples/$entity",
      "Address": null,
      "Age": 18,
      "ID": testPeopleID,
      "Name": "theo valid",
      "Weight": 9999,
    });
  });

  it("should support update the Age of people", async () => {
    const response = await axios.patch(`/people/EarthPeoples(${testPeopleID})`, { Age: 99 });
    expect(response.status).toBe(200);
    expect(response.data).toMatchObject({
      ID: testPeopleID,
      Age: 99
    });
  });

  it("should support soft delete for EarthPeople", async () => {
    const response = await axios.delete(`/people/EarthPeoples(${testPeopleID})`);
    expect(response.status).toBe(204);
    // because setup the spy on the cds.db.run
    // so that, you can check the last call of the cds.db.run
    expect(cds.db.run.mock.lastCall).toMatchSnapshot();
    // verify
    const deletedQuery = await axios.get(`/people/EarthPeoples(${testPeopleID})`);
    expect(deletedQuery.status).toBe(404);
    const dbPeople = await cds.run(SELECT.one.from("db.EarthPeople", testPeopleID));
    expect(dbPeople).not.toBeNull();
  });

  it("should support cds API and validation", async () => {
    // you can use cds.xxx api here, and test framework directly
    const peopleService = await cds.connect.to("PeopleService");
    // you want to expect 'async' error
    await expect(async () => {
      await peopleService.run(INSERT.into("EarthPeoples").entries({
        Name: "Theo"
      }));
    }).rejects.toThrow("invalid name");
  });

  it("should support raise error when delete without query filter", async () => {
    const peopleService = await cds.connect.to("PeopleService");
    await expect(() => peopleService.run(DELETE.from("EarthPeoples")))
      .rejects
      .toThrow("internal server error");
  });

  it("should support snapshot test", async () => {
    const response = await axios.post("/people/EarthPeoples", {
      ID: "18f3dfbf-92e2-4e50-a14d-ca120fb3aff5",
      Name: "Theo for Snapshot"
    });
    expect(response.status).toBe(201);
    expect(response.data).toMatchSnapshot();
  });

});


