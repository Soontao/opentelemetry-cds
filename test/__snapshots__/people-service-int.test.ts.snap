// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`People Service Int Test should support snapshot test 1`] = `
{
  "@odata.context": "$metadata#EarthPeoples/$entity",
  "Address": null,
  "Age": 18,
  "ID": "18f3dfbf-92e2-4e50-a14d-ca120fb3aff5",
  "Name": "Theo for Snapshot",
  "Weight": 9999,
}
`;

exports[`People Service Int Test should support soft delete for EarthPeople 1`] = `
[
  Query {
    "UPDATE": {
      "data": {
        "Deleted": true,
        "ID": "318da8b4-95be-498d-bae7-f0c6ed6516ac",
      },
      "entity": "db.EarthPeople",
      "where": [
        {
          "ref": [
            "ID",
          ],
        },
        "=",
        {
          "val": "318da8b4-95be-498d-bae7-f0c6ed6516ac",
        },
      ],
    },
  },
]
`;
